docs/vm-configuration.md

```diff
@@ -266,3 +266,227 @@ the generation of multiple libvirt XML documents, the domain XML and associated
 documents for secrets, network filters and node devices. So there is not a strict
 1-1 mapping between the k8s API resource representation and the libvirt low
 level representation.
+
+Mapping rules for mapping libvirt XML to the KubeVirt API
+---------------------------------------------------------
+
+In order to support libvirt features, we tend to expose the libvirt API as much
+as possible in a 1:1 fashion. However, there are some differences which need to
+be taken into account:
+
+ 1. Our API is json/yaml and not xml.
+ 2. Host specifics are not exposed on the API.
+ 3. We want to make use of the golang xml/yaml/json mappers as much as possible.
+ 4. Ordering of devices can be important (e.g. for pci address assignment).
+ 5. KubeVirt also delivers specialized device types (mostly for network and
+    storage), which need to be fit into the existing structure.
+
+**Note**: Point **4** is still under debate.
+
+## General mapping rules
+
+### All devices need to live directly under *devices*
+
+In order to allow arbitrary ordering of devices, they all need directly live
+under `devices`.
+
+This is OK:
+
+```yaml
+devices:
+  lun:
+    my: data
+  video:
+    qxl:
+      my: data
+  lun:
+    my: data1
+  video:
+    qxl:
+      my: data1
+```
+
+This is **not** OK:
+
+```yaml
+devices:
+  disks:
+    lun:
+      my: data
+    lun:
+      my: data1
+  videos:
+    qxl:
+      my: data
+    qxl:
+      my: data1
+```
+
+### Add clear extension points where KubeVirt provides specializations
+
+Disks and network interfaces are good examples where KubeVirt provides
+specializations. See the mapping decisions for them below to get an impression
+on how clear extension points can be added. In general allowing to choose
+between different optional source or target structs is the preferred mechanism,
+to cleanly separate between KubeVirt extensions and libvirt source or target
+types.
```

Comment:
I don't understand what you mean with the separation here. From the examples below it seems to me that KubeVirt extensions and libvirt structures are seamlessly integrated. Could you please clarify the intent here?

Reply:
IIUIC then yes, from an API point of view extensions and libvirt support are not distinguishable.
But by using this API approach they can be cleanly separated on the implementation side.

Reply:
@fabiand exactly.

It is a difference, if I have one big json blob which contains all fields, vs. having different optional subsections, where only one is allowed to be filled out.

I think the difference comes obvious immediately when e.g. looking at the VolumeSource implementation in k8s: https://github.com/kubernetes/kubernetes/blob/029ed12b332dc358782a453f9ed904fa16844d8c/staging/src/k8s.io/api/core/v1/types.go#L250

It would be extremely hard to distinguish between different volume sources, if all fields for all types would be one blob in the VolumeSource struct.